Implemented dependsOn/dependencyOf support in catalog model.

jmgrimes opened the pull request:
Added support for the dependsOn property to the Component kind in the catalog model and schema, plus tests.
Added support for the dependencyOf property to the Resource kind in the catalog model and schema, plus tests.
Added support for generating relationships from these new properties via the BuiltinKindsEntityProcessor, plus tests.
Added support for fetching dependsOn relationships between components in the system to render the System Diagram.

Closes #4505

freben commented:
Hi! Just a quick note that this can be rebased on latest master now - i see the diff has a number of unrelated changes in it at the moment.

Also the changesets can bet only `patch` level of the changes are backwards compatible additions.

We will review more closely after the weekend :) thanks for making this!

jmgrimes commented:
Ok perfect!  Saw the help wanted tag on this one, and thought I understood it well enough to take a shot at it.

Rebased to latest master, and updated changeset for all packages/plugins to patch level.

adamdmharvey commented:
Closes https://github.com/backstage/backstage/issues/4505

benjdlambert commented:
I think this looks pretty good, can I get a review from @Rugvip / @freben too :)

benjdlambert commented on .changeset/red-planets-hammer.md:
```diff
@@ -0,0 +1,7 @@
+'@backstage/plugin-catalog': patch
+---
+
+Implemented missing support for the dependsOn/dependencyOf relationships between Component and Resource catalog model objects. Added support for generating the relevant relationships to the BuiltinKindsEntityProcessor, and added simple support for fetching relationships between Components and Resources for rendering in the system diagram. All catalog-model changes backwards compatible.
```
Think we can maybe fix the formatting here a little bit?

Maybe things like BuiltinKindsEntityProcessor could be `BuiltinKindsEntityProcessor` and some line breaks to make the change log easier to read :)

jmgrimes replied:
Added the formatting as requested.

freben commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -519,6 +519,15 @@ consumed by the component, e.g. `artist-api`. This field is optional.
+
+| [`kind`](#apiversion-and-kind-required) | Default [`namespace`](#namespace-optional) | Generated [relation](well-known-relations.md) type                                            |
+| --------------------------------------- | ------------------------------------------ | --------------------------------------------------------------------------------------------- |
+| [`Resource`](#kind-resource) (default)  | Same as this entity, typically `default`   | [`dependsOn`, and reverse `dependencyOf`](well-known-relations.md#dependson-and-dependencyof) |
```
This is correct in terms of the diagram - but I'm thinking that we should permit dependencies at least on `Component` kinds too.

On the one hand, we could postpone that and get this great change in without blocking too much.

On the other hand, the main risk with that is that we might want to NOT have a default here if there's no natural best default. And removing the default would be a breaking change, making people's YAML files break.

So ...

Maybe best to already now make both `Component` and `Resource` supported, and have no default?

@backstage/maintainers @Fox32 @andrewthauer @adamdmharvey @dtuite @dhenneke as people of great wisdom and customer insight - how do you feel about `Component dependsOn Component`? There's more stuff than just APIs that one can need for one's functioning - are all those other stuff `Resource`s?

Fox32 replied:
`Component dependsOn Component` is something I would like to see in the long term. Could be suited to show dependencies between libraries/packages. We could parse an NPM registry/maven repo/you name it and show which components depend on each other. I wouldn't implement it now though.

If I think about the default, I can't decide whether `Component` or `Resource` should be the default. So let's keep it open for now. Adding a default later would be a non breaking change I guess.

freben replied:
Agreed. So the vote is to not intentionally limit the kinds here, and also to have no default right

andrewthauer replied:
I'm in full agreement of supporting `Component dependsOn Component` as well. In some cases it might be more common to use that relation then `Resource` since a component would likely be something you own and maintain vs a resource is typically an external concern (e.g. cloud service).

> We could parse an NPM registry/maven repo/you name it and show which components depend on each other. I wouldn't implement it now though.

We've thought about doing this as well and agree it would be very useful. Just thinking out loud but ... how would this work for 3rd party libraries? Would those also be Components that are depended on by an internal Component? Or possibly a relation to a new entity type that is not necessarily tracked directly in the catalog. Maybe something else.

freben replied:
Yeah i've been pondering concerns like that as well. To express relations to things, we now _have_ to import them. What if the catalog rather could express relations and metadata on top of essentially "foreign keys" - so that you could for example say that you depend on `<URN or URL expressing a specific NPM package>`, instead of just entity refs? I dunno, it's just a flicker of an idea in my mind.

jmgrimes replied:
What about taking the approach of using a separate field for each of these types?  Perhaps "dependsOnResources" and "dependsOnComponents" or some similar type of set of fields to differentiate this?  This would be similar to the approach used by the `Group` entity when specifying child groups and direct members (requested in #5072) with "children" and "members" fields in the spec.  It also seems like this implies the need to support "dependencyOf" for other components here, perhaps with a third "dependencyOfComponents" field. Thoughts?

Otherwise, I'm not able to find any examples of non-defaulted kinds when calling `doEmit` in the existing codebase, nor any implementation that restricts the entity reference kinds that can be passed.  In fact, digging a little bit, it seems like each of these entity reference fields would reasonably accept any entity type passed, even if the reference didn't make sense.  Could anyone point me in the direction of this enforcement if it does exist, if we prefer this to be through a single field?

Rugvip replied:
We do also have the naming rules laid out next to the definitions of the relations: https://github.com/backstage/backstage/blob/c36a7847e49a1b61f76cff4451d30461c96e1246/packages/catalog-model/src/kinds/relations.ts#L18-L22

freben replied:
And yes, just skip the restrictions in code. You can just emit what the component spec declares.

jmgrimes replied:
Sounds good, I'll proceed without the restrictions in code and keep things as using the "dependsOn" and "dependencyOf" semantics.  Thanks!

freben commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -1001,6 +1010,15 @@ belongs to, e.g. `artist-engagement-portal`. This field is optional.
+
+| [`kind`](#apiversion-and-kind-required)  | Default [`namespace`](#namespace-optional) | Generated [relation](well-known-relations.md) type                                            |
+| ---------------------------------------- | ------------------------------------------ | --------------------------------------------------------------------------------------------- |
+| [`Component`](#kind-component) (default) | Same as this entity, typically `default`   | [`dependencyOf`, and reverse `dependsOn`](well-known-relations.md#dependson-and-dependencyof) |
```
Open question here too, should we really open up for the reverse dependency in the spec? It feels like a pretty fringe thing.

Fox32 replied:
We could argue that it we do it for groups and users too, but on first sign I would say that my component references something. I think we should be flexible.

freben replied:
Ok so vote to keep this, also with no default?

But the tables (here and above) should list both Resource and Component as alternatives I think

jmgrimes replied:
I think in this case, from the Resource direction outward, the only dependencyOf relation that might make sense would be a `Component` though, right?

freben replied:
Not sure! This is one of the reasons this wasn't in place yet I guess, it's a little hazy still what the details should be :)

Again, probably we'll have to not restrict, and leave without a default.

jmgrimes replied:
I'm still that adding `Resource` here wouldn't be semantically appropriate, as it would allow the declaration of `Resource` dependencyOf `Resource`.  I don't know if that's useful or not?

I'm tending to agree with you that this side of the declaration of the relationship might not be all that useful.  Much more reasonable to say that my `Component` dependsOn `Resource` which was the original intent of the Issue.

freben commented:
@jmgrimes yeah this is great stuff, just a tweak to what's suggested and removing the default, and this will be ready to go!

jmgrimes commented:
I'm having a number of second thoughts about this as I try to implement the non-default options.  Briefly, these are:

* The reverse dependencyOf relationship adds some flexibility, but we'd need to add dependencyOf support to `Component`.  We'd also need to add the dependsOn field to the `Resource` type to fully support this reverse declaration.  I'm not sure this makes sense, as it'd be uncommon for an external resource to dependOn anything in the ecosystem.
* Declaring these types of flexible relationships let us do some strange things like `Component A` dependsOn `Resource B` dependsOn `Component C` dependsOn `Component A`.  This creates cycles in an otherwise clean, top down dependency graph.  Taking reverse dependencyOf out fixes this.
* A `Component` could be a dependency and a subcomponent.  This could create a similar situation as above.  Maybe this isn't a big deal, but it is a possibly unintended side effect.
* The existing `doEmit` method requires a defaultKind to be defined in the context.  Changing this signature causes the arguments. to `result.relation` to no longer be valid.  I could leave in the defaults in the code and remove them from the documentation possibly?  Or I could attempt some refactoring.

I propose:

* Remove the dependencyOf field from the `Resource` entity.  The reverse relationship doesn't feel useful and could be very confusing for users.
* Remove the default entity kind for the dependsOn field in the documentation, listing support for dependsOn `Component` and `Resource`.
* Keep the default in the `BuiltinKindsEntityProcessor` for the dependsOn field as `Resource` to avoid significant refactor.  The documentation and examples will guide users to the proper usage.

I think this adds the flexibility of the `Component` dependsOn `Component` relationship,  to introduce cycles or get mired in too many options for this declaration.

Thoughts @freben, @Fox32, @benjdlambert?

Fox32 commented:
> This creates cycles in an otherwise clean, top down dependency graph. 

Hmm, we try to be able to model real world software, right? Even though I don't think that cycling dependencies are good, they still often occur.

freben commented:
Yep we can't assume no cycles from a code standpoint at least. Also remember that all relations are bidirectional, so there's little circular pairs all over the place :)

freben commented:
> * The reverse dependencyOf relationship adds some flexibility, but we'd need to add dependencyOf support to `Component`.  We'd also need to add the dependsOn field to the `Resource` type to fully support this reverse declaration.  I'm not sure this makes sense, as it'd be uncommon for an external resource to dependOn anything in the ecosystem.

Not sure we can consider a Resource as being "external". I would say that the resource could just as well represent memcached sidecars running on the exact same pods as the service. We don't know yet all the ways that users will make use of the (kinda loose and open) system model.

Also reiterating: the relations are always bidirectional between pairs of entities, but this does _not_ mean that we have to make the yaml fields to generate them bidirectional, or even exist at all. For example APIs yaml files are _not_ able to say what consumers they have, and I don't think they should either. Some future relations may even turn out to be entirely generated "implicitly" without user interaction.

Given the above, we are kinda slowly and carefully adding things to the yaml format for generating relations. We thought that the idea of `dependsOn` passed the threshold for maybe being time to introduce, even though it's a quirky one. And I vote that we'll just completely ignore making the reverse one for now because it doesn't feel comfortable, possibly for good reason.

> * Declaring these types of flexible relationships let us do some strange things like `Component A` dependsOn `Resource B` dependsOn `Component C` dependsOn `Component A`.  This creates cycles in an otherwise clean, top down dependency graph.  Taking reverse dependencyOf out fixes this.

Not if we allow

```yaml
kind: Component
metadata:
  name: c
spec
  dependsOn: Resource:r
---
kind: Resource
metadata:
  name: r
spec
  dependsOn: Component:c
```

Which I think we should.

> * A `Component` could be a dependency and a subcomponent.  This could create a similar situation as above.  Maybe this isn't a big deal, but it is a possibly unintended side effect.

Not really unintended, the relations express different statements about how those two inter-relate. It may not make a lot of sense to make those two claims typically, but we don't know yet.

> * The existing `doEmit` method requires a defaultKind to be defined in the context.  Changing this signature causes the arguments. to `result.relation` to no longer be valid.  I could leave in the defaults in the code and remove them from the documentation possibly?  Or I could attempt some refactoring.

Just make the default kind optional, and solve the code inside the method accordingly, making sure that either the string ref or the default kind had some value in them. If none were given, throw an error.

> I propose:
> 
> * Remove the dependencyOf field from the `Resource` entity.  The reverse relationship doesn't feel useful and could be very confusing for users.

I think neither `Component` nor `Resource` should have the `dependencyOf` _spec field_ at this moment, but they will get the generated `dependencyOf` _relation_ generated either way. Only thing to add is the `dependsOn` _spec field_ to both `Component` and `Resource`.

> * Remove the default entity kind for the dependsOn field in the documentation, listing support for dependsOn `Component` and `Resource`.

Yep

> * Keep the default in the `BuiltinKindsEntityProcessor` for the dependsOn field as `Resource` to avoid significant refactor.  The documentation and examples will guide users to the proper usage.

I still think this emit helper can be safely refactored instead.

> I think this adds the flexibility of the `Component` dependsOn `Component` relationship, to introduce cycles or get mired in too many options for this declaration.

Yeah I think we've arrived at having to support cyles.

jmgrimes commented:
Perfect.  Thanks for helping me work through this one in my head a bit better.  On it!

freben commented:
I wanna say I really enjoy the discourse. It's clear that you're giving this careful thought and it's appreciated. I hope we don't come off as bullish or over-opinionated or so :) Thanks for your contributions.

jmgrimes commented:
Absolutely!  I'm really excited about this particular project, and am hoping to adopt this internally, since we've built a portal for a few of our apps, but nothing as complete or comprehensive as this.  A particular thing I needed recently was the ability to do some high level system diagramming for which this feature adds a lot of lift for that use case.  Just wanted to be sure I was doing the right thing, and love the discussion.  Please let me know if I can be of further assistance!

freben commented on packages/catalog-model/src/schema/kinds/Component.v1alpha1.schema.json:
```diff
@@ -84,6 +84,14 @@
+            },
+            "dependsOn": {
+              "type": "array",
+              "description": "An array of entity references to the resources that the component depends on.",
```
```suggestion
              "description": "An array of references to other entities that the component depends on to function.",
```

freben commented on packages/catalog-model/src/schema/kinds/Resource.v1alpha1.schema.json:
```diff
@@ -47,6 +47,14 @@
             },
+            "dependsOn": {
+              "type": "array",
+              "description": "An array of entity references to the components that the resource is a dependency of.",
```
```suggestion
              "description": "An array of references to other entities that the resource depends on to function.",
```

freben commented on plugins/catalog-backend/src/ingestion/processors/BuiltinKindsEntityProcessor.ts:
```diff
@@ -99,16 +101,29 @@ export class BuiltinKindsEntityProcessor implements CatalogProcessor {
         const targetRef = parseEntityRef(target, context);
+        if (targetRef.kind === undefined) {
+          throw new Error(
+            'Entity reference kind is undefined and has no default',
```
```suggestion
            `Entity reference "${target}" did not specify a kind (e.g. starting with "Component:"), and has no default`,
```

we could also give a `fieldName` param to this function or something, so that we can say in the error message which field that had the error, to be extra super-duper-helpful

freben replied:
(you don't have to do the fieldName thing, that was just reflection on future extension of this)

freben reviewed (commented):
Cool! A few minor things, then good to go!

jmgrimes commented:
@freben Anything else you'd like to see implemented on this?  I thought about the field name side for error handling, but thought maybe that would be best as a separate issue?

Rugvip commented:
@jmgrimes Just a heads up that freben is on vacation until next week, and I'll let him finish this off as he has all the context

freben reviewed (approved):
Woo!
